iroh/src/baomap/flat.rs

```diff
@@ -0,0 +1,1210 @@
+            Either::Left(mem) => mem.len() as u64,
+            Either::Right((_, size)) => *size,
+        }
+    }
+}
+
+fn needs_outboard(size: u64) -> bool {
+    size > (IROH_BLOCK_SIZE.bytes() as u64)
+}
+
+impl Map for Store {
+    type Entry = Entry;
+    type Outboard = PreOrderOutboard<MemOrFile>;
+    type DataReader = MemOrFile;
+    fn get(&self, hash: &Hash) -> Option<Self::Entry> {
+        let state = self.0.state.read().unwrap();
+        if let Some(entry) = state.complete.get(hash) {
+            tracing::trace!("got complete: {} {}", hash, entry.size);
+            let outboard = state.load_outboard(entry.size, hash)?;
+            // check if we have the data cached
+            let data = state.data.get(hash).cloned();
+            Some(Entry {
+                hash: blake3::Hash::from(*hash),
+                entry: EntryData {
+                    data: if let Some(data) = data {
+                        Either::Left(data)
+                    } else {
+                        // get the data path
+                        let path = if entry.owned_data {
+                            // use the path for the data in the default location
+                            self.owned_data_path(hash)
+                        } else {
+                            // use the first external path. if we don't have any
+                            // we don't have a valid entry
+                            entry.external_path()?.clone()
+                        };
+                        Either::Right((path, entry.size))
+                    },
+                    outboard: Either::Left(outboard),
+                },
+            })
+        } else if let Some(entry) = state.partial.get(hash) {
+            let data_path = self.0.options.partial_data_path(*hash, &entry.uuid);
+            let outboard_path = self.0.options.partial_outboard_path(*hash, &entry.uuid);
+            tracing::trace!(
+                "got partial: {} {} {}",
+                hash,
+                entry.size,
+                hex::encode(entry.uuid)
+            );
+            Some(Entry {
+                hash: blake3::Hash::from(*hash),
+                entry: EntryData {
+                    data: Either::Right((data_path, entry.size)),
+                    outboard: Either::Right(outboard_path),
+                },
+            })
+        } else {
+            tracing::trace!("got none {}", hash);
+            None
+        }
+    }
+}
+
+impl ReadonlyStore for Store {
+    fn blobs(&self) -> Box<dyn Iterator<Item = Hash> + Send + Sync + 'static> {
+        let inner = self.0.state.read().unwrap();
+        let items = inner.complete.keys().copied().collect::<Vec<_>>();
+        Box::new(items.into_iter())
+    }
+
+    fn roots(&self) -> Box<dyn Iterator<Item = Hash> + Send + Sync + 'static> {
+        unimplemented!()
+    }
+
+    fn validate(&self, _tx: mpsc::Sender<ValidateProgress>) -> BoxFuture<'_, anyhow::Result<()>> {
+        unimplemented!()
```

Comment:
what about this?

Reply:
Had punted on this, see https://github.com/n0-computer/iroh/issues/1338

I guess an anyhow::bail might be nicer.